smithy-aws-endpoints/src/main/java/software/amazon/smithy/rulesengine/aws/validators/EndpointModifierValidator.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.rulesengine.aws.validators;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import software.amazon.smithy.model.FromSourceLocation;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.rulesengine.aws.traits.StandardPartitionalEndpointsTrait;
+import software.amazon.smithy.rulesengine.aws.traits.StandardRegionalEndpointsTrait;
+import software.amazon.smithy.utils.SetUtils;
+
+/**
+ * Validate EndpointModifier traits that are applied on a service.
+ */
+public class EndpointModifierValidator extends AbstractValidator {
+
+    private static final Set<String> SUPPORTED_PATTERNS = SetUtils.of(
+            "{region}", "{service}", "{dnsSuffix}", "{dualStackDnsSuffix}"
+    );
+
+    private static final Pattern PATTERN = Pattern.compile("\\{[^\\}]*\\}");
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+        for (ServiceShape serviceShape : model.getServiceShapesWithTrait(StandardRegionalEndpointsTrait.class)) {
+            events.addAll(validateRegionalEndpointPatterns(serviceShape,
+                    serviceShape.expectTrait(StandardRegionalEndpointsTrait.class)));
+        }
+        for (ServiceShape serviceShape : model.getServiceShapesWithTrait(StandardPartitionalEndpointsTrait.class)) {
+            events.addAll(validatePartitionalEndpointPatterns(serviceShape,
+                    serviceShape.expectTrait(StandardPartitionalEndpointsTrait.class)));
+        }
+        return events;
+    }
+
+    private List<ValidationEvent> validateRegionalEndpointPatterns(
+            ServiceShape serviceShape, StandardRegionalEndpointsTrait regionalEndpoints) {
+        List<ValidationEvent> events = new ArrayList<>();
+
+        regionalEndpoints.getRegionSpecialCases().forEach((region, specialCases) -> {
```

Comment:
Since `region` looks unused, you could iterate over the `values()` instead of the whole map.

```
for (RegionSpecialCase specialCase : regionalEndpoints.getRegionSpecialCases().values())
```

nit: also prefer `for` loops to `forEach` streaming iteration.

Reply:
This feedback applies to the `Partitional` branch as well, with the naming replaced.